### PR TITLE
Allow parsing `ref readonly` modifier

### PR DESCRIPTION
**Pull request by jjonescz**
Speclet: https://github.com/dotnet/csharplang/blob/0c269d235e0687419c1ec69743fb51cea8b4d466/proposals/ref-readonly-parameters.md
Test plan: https://github.com/dotnet/roslyn/issues/68056

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:617]: It looks like `readonly` is allowed in combination with `this`. Can we add a test?

**Review by jcouv — COMMENTED** [review]
It looks like an alternative syntax is still being discussed (using an attribute), so I'll hold off review until we close on that

File: src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs:705]: Consider also testing with `scoped`

**Review by jcouv — DISMISSED** [review]
LGTM Thanks (iteration 1) with some small test suggestions

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (Sergio0694) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:617]: Should we also add a test to double check that invoking `ref readonly` extension methods by pointers also works?
That is, this should work (just like it does for `ref` and `in` extension methods invoked by pointers):

```csharp
public static class Ex
{
    public static void Foo(this ref readonly int x)
    {
    }

    static void Test()
    {
        int x = 42;
        (&x)->Foo();
    }
}
```

Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:617]: Thanks @Sergio0694. This PR is only about syntax, but will add a note to test plan.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs
Comment (Sergio0694) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs:803]: Should this actually be allowed? 🤔
If we said that the `readonly` position after `ref` should be strict, then shouldn't `ref readonly this` be the only other allowed combination other than `this ref readonly`, whereas `ref this readonly` should be disallowed?

Comment (jjonescz) [reply src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs:803]: Makes sense, thanks. My initial thinking was that `this` is allowed to travel anywhere, but I guess it's better to be as strict as possible for future flexibility. (And actually `this` cannot even be specified after the modifiers if there's `scoped`.)

**Comment by jjonescz** [comment]
@jcouv, @AlekseyTs PTAL

**Comment by jjonescz** [comment]
@jcouv @AlekseyTs @dotnet/roslyn-compiler please review

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:4449]: Do we need a similar change in `ParsePossibleScopedKeyword`?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 7)

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:617]: >seenReadonly = true; [](http://example.com/codeflow?start=28&length=20)

Can we assert that ```seenRef``` is true here?

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:500]: >addERR_BadParameterModifiers(diagnostics, modifier, SyntaxKind.ReadOnlyKeyword); [](http://example.com/codeflow?start=28&length=80)

Is this code path reachable?

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:535]: >addERR_BadParameterModifiers(diagnostics, modifier, SyntaxKind.ReadOnlyKeyword); [](http://example.com/codeflow?start=28&length=80)

Is this code path reachable?

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:574]: >addERR_BadParameterModifiers(diagnostics, modifier, SyntaxKind.ReadOnlyKeyword); [](http://example.com/codeflow?start=28&length=80)

Is this code path reachable?

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:593]: >Binder.CheckFeatureAvailability(modifier, MessageID.IDS_FeatureRefReadonlyParameters, diagnostics); [](http://example.com/codeflow?start=24&length=99)

Consider performing this check only if ```seenReadonly``` is false.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:610]: It looks like these checks are essentially duplicates of the checks performed for a `ref` modifier, and below we report an error if previous modifier is not `ref`. Therefore, consider simplifying the logic and removing these checks, they are going to produce more "noise". At the same time, I think it is more important to report  ```ERR_RefReadOnlyWrongOrdering``` rather than the "noise".

File: src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs:544]: >RefReadonly_Compilation [](http://example.com/codeflow?start=20&length=23)

Semantic diagnostics tests do not belong in parsing tests. Please move them to the file that will be responsible for testing semantics for `ref readonly` parameters. Some legacy tests do that because they used to test for parsing diagnostic, but then the phase reporting the diagnostic got changed and the tests were adjusted in place.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs:544]: I recommend creating a new file in Emit2 project. We are pretty close to reaching the string heap limit in older test projects.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs:999]: >ref scoped readonly int p [](http://example.com/codeflow?start=41&length=25)

It would be good to assert nodes for this code.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs:1045]: >readonly scoped ref int p [](http://example.com/codeflow?start=41&length=25)

It would be good to assert nodes for this code.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs:1091]: >scoped readonly int p [](http://example.com/codeflow?start=41&length=21)

It would be good to assert nodes for this code.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Syntax/Parsing/RefReadonlyTests.cs:1091]: Consider also testing nodes for ```scoped readonly ref int p```

**Comment by AlekseyTs** [comment]
Done with review pass (commit 7)

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:535]: Yes, in `void M(readonly params int[] p)`.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:535]: > Yes, in ```void M(readonly params int[] p)```.

I might be missing something, but in ```ReadonlyWithParams``` unit-test I do not see this error reported.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:535]: In fact, I do not see ERR_BadParameterModifiers reported with "'readonly'" at the end in the entire file.

Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs]: Yes, sorry, posted the reply by mistake.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs]: Done, but it doesn't have much effect as `seenReadonly` is `false` when there are other errors (like missing `ref`).

Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs]: Instead, I can move the check where we set `seenReadonly` to `true` to reduce diagnostics even further, see ee2b4370b67.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:478]: >Debug.Assert(!seenReadonly); [](http://example.com/codeflow?start=24&length=28)

I am not sure why this should be the case. We can assert that ```seenReadonly``` is false unless ```seenRef``` is true, but I do not think we can assume that  ```seenReadonly``` is always false here. We probably should add a test that demonstrates that.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:507]: >Debug.Assert(!seenReadonly); [](http://example.com/codeflow?start=24&length=28)

Same comment here

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:545]: >Debug.Assert(!seenReadonly); [](http://example.com/codeflow?start=24&length=28)

Same comment here

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:224]: >// (3,23): error CS9501: 'readonly' modifier must be specified after 'ref'. [](http://example.com/codeflow?start=16&length=75)

This error feels misleading, `readonly` is specified after `ref` in code. I think we should suppress this error.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:585]: >else if (!seenRef || previousModifier?.Kind() != SyntaxKind.RefKeyword) [](http://example.com/codeflow?start=24&length=71)

This is related to the comment for ```RefReadonlyWithParams_01``` unit-test. We probably shouldn't check ```!seenRef``` here. Also, we probably should change the following ```else``` to ```else if (seenRef)``` and remove the assert in its the block.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 11), assuming CI is passing

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:8]: Consider using file-scoped namespace (we tend to do that for new test files)

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:589]: Do we need this condition?

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:592]: Should we unconditionally set `seenReadonly`?

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:8]: > The main reason I advocate this is because test files tend to have long lines (such as diagnostics).

I dislike this form of namespaces no matter what.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:589]: > Do we need this condition?

It makes an observable difference and, in my opinion, improves quality of diagnostics.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:592]: What is the reason to set `seenReadonly` conditionally?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:592]: > What is the reason to set seenReadonly conditionally?

It does not look useful to set it in a any other condition. Other cases do not look at the value. An error that `readonly` it is not preceded by `ref` is much better than an error about a duplicate, in my opinion.

Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:592]: FWIW, this is the difference: https://github.com/dotnet/roslyn/commit/8cdd5410f44aa11cccb2e87225be24f1d89a6e36

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:592]: Basically `readonly` means nothing unless it is part of `ref readonly` and the ref is actually not ignored.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:589]: It looks like one test would be affected: `params ref readonly`. I think you're correct, this check reduces cascaded diagnostics in such scenario.
Then my problem is with how the variable is named, it would be better as "seenValidRef". That said, previous variables seem to suffer the same problem, so it's fine to leave.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 11)

**Pull request merged**
